In [3]:
# import Pkg; Pkg.add("CSV")
# import Pkg; Pkg.add("Grep")
# import Pkg; Pkg.add("FixedEffectModels")

In [4]:
using CSV, DataFrames, Grep
using RData, LinearAlgebra, GLM, DataFrames, Statistics, Random, Distributions, DataStructures, NamedArrays, PrettyTables
import CodecBzip2


This notebook contains an example for teaching.

# A Case Study: The Effect of Gun Ownership on Gun-Homicide Rates

We consider the problem of estimating the effect of gun
ownership on the homicide rate. For this purpose, we estimate the following partially
linear model

$$
 Y_{j,t} = \beta D_{j,(t-1)} + g(Z_{j,t}) + \epsilon_{j,t}.
$$

## Data

$Y_{j,t}$ is log homicide rate in county $j$ at time $t$, $D_{j, t-1}$ is log  fraction of suicides committed with a firearm in county $j$ at time $t-1$, which we use as a proxy for gun ownership,  and  $Z_{j,t}$ is a set of demographic and economic characteristics of county $j$ at time $t$. The parameter $\beta$ is the effect of gun ownership on the
homicide rates, controlling for county-level demographic and economic characteristics. 

The sample covers 195 large United States counties between the years 1980 through 1999, giving us 3900 observations.

In [646]:
data = CSV.File("../data/gun_clean.csv") |> DataFrame
size(data)

(3900, 415)

### Preprocessing

To account for heterogeneity across counties and time trends in  all variables, we remove from them county-specific and time-specific effects in the following preprocessing.

In [6]:
# #################################  Find Variable Names from Dataset ########################

# varlist <- function (df=NULL,type=c("numeric","factor","character"), pattern="", exclude=NULL) {
#   vars <- character(0)
#   if (any(type %in% "numeric")) {
#     vars <- c(vars,names(df)[sapply(df,is.numeric)])
#   }
#   if (any(type %in% "factor")) {
#     vars <- c(vars,names(df)[sapply(df,is.factor)])
#   }  
#   if (any(type %in% "character")) {
#     vars <- c(vars,names(df)[sapply(df,is.character)])
#   }  
#   vars[(!vars %in% exclude) & grepl(vars,pattern=pattern)]
# }

In [7]:
################################# Create Variables ###############################


# Dummy Variables for Year and County Fixed Effects
R = r"X_Jfips"
fixed = grep(R, names(data))
year = grep("X_Tyear", names(data))
println("Variable: <<Fixed>> has ", length(fixed), " features")
println("Variable: <<Year>> has ", length(year), " outcomes")

# Census Control Variables

census = []
census_var = ["AGE", "BN", "BP", "BZ", "ED", "EL", "HI", "HS", "INC", "LF", "LN", "PI", "PO", "PP", "PV", "SPR", "VS"]

for i in 1:size(census_var, 1) 
    append!(census, grep(census_var[i], names(data)))
end

println("Variable: <<Census>> has ", length(census), " features")


Variable: <<Fixed>> has 195 features
Variable: <<Year>> has 21 outcomes
Variable: <<Census>> has 186 features


In [8]:
################################ Variables ##################################
# Treatment Variable
d = "logfssl"

# Outcome Variable
y = "logghomr"

# Other Control Variables
X1 = ["logrobr", "logburg", "burg_missing", "robrate_missing"]
X2 = ["newblack", "newfhh", "newmove", "newdens", "newmal"]



#################################  Partial out Fixed Effects ########################

# New Dataset for Partiled-out Variables
rdata = DataFrame(CountyCode = data[:,"CountyCode"])

# Variables to be Partialled-out
varlist2 = vcat(y, d, X1, X2, census)
println("Variable: <<varlist2>> has ", length(varlist2), " features")

form = []

# Partial out Variables in varlist from year and county fixed effect
for i in 1:size(varlist2, 1)
    append!(form, [term.(varlist2[i]) ~ sum(term.(vcat(year, fixed)))])
    rdata[!, varlist2[i]] = residuals(lm(form[i], data))
end

rdata

Variable: <<varlist2>> has 197 features


3900×198 DataFrame
  Row │ CountyCode  logghomr    logfssl      logrobr      logburg      burg_mi ⋯
      │ Int64       Float64     Float64      Float64      Float64      Float64 ⋯
──────┼─────────────────────────────────────────────────────────────────────────
    1 │       1073  -0.134778    0.0961271    0.150893    -0.124395      0.010 ⋯
    2 │       1073  -0.239622    0.0808094    0.0401683   -0.134781      0.010
    3 │       1073  -0.0786772   0.0573399   -0.017679    -0.167909      0.010
    4 │       1073  -0.331465    0.0816945   -0.00963344  -0.22925       0.010
    5 │       1073  -0.31664     0.0253655   -0.0267151   -0.176635      0.003 ⋯
    6 │       1073   0.105132   -0.00677726  -0.151487    -0.189069      0.010
    7 │       1073  -0.0373401   0.0773061   -0.166729    -0.117739      0.010
    8 │       1073  -0.0520609  -0.108433    -0.0996453   -0.0833094     0.004
    9 │       1073   0.0547007  -0.0340988    0.151557     0.319282     -0.044 ⋯
   10 │       1073   0.122094   -0.0824292    0.0476034   -0.0144728    -0.002
   11 │       1073   0.110314   -0.0243444    0.00814297  -0.0349694     0.010
  ⋮   │     ⋮           ⋮            ⋮            ⋮            ⋮            ⋮  ⋱
 3891 │      55133  -0.54942    -0.00681523   0.0786474    0.13458      -0.020
 3892 │      55133  -0.660408    0.135748     0.0561803    0.134567     -0.021 ⋯
 3893 │      55133  -0.676525   -0.0223923    0.209323     0.390146     -0.060
 3894 │      55133  -0.697525    0.0405164   -0.0858534    0.15756      -0.021
 3895 │      55133  -0.0384095  -0.058838    -0.0983204    0.126737     -0.023
 3896 │      55133   0.0280962   0.0407888    0.0182247    0.107714     -0.030 ⋯
 3897 │      55133   0.0801447   0.208251     0.251402     0.151313     -0.030
 3898 │      55133   0.497987    0.177444     0.179985     0.116497     -0.023
 3899 │      55133   0.568542   -0.0971269   -1.3754      -3.02227       0.530
 3900 │      55133   0.609022    0.274062     0.336788     0.171876     -0.036 ⋯
                                               193 columns and 3879 rows omitted

In [9]:
# load dataset
rdata_read = CSV.File("../data/gun_clean2.csv") |> DataFrame
data = rdata_read[!, names(rdata)]
n = size(data,1)

3900

In [10]:
column_names = names(data)

198-element Vector{String}:
 "CountyCode"
 "logghomr"
 "logfssl"
 "logrobr"
 "logburg"
 "burg_missing"
 "robrate_missing"
 "newblack"
 "newfhh"
 "newmove"
 "newdens"
 "newmal"
 "AGE010D"
 ⋮
 "PVY020D"
 "PVY120D"
 "PVY210D"
 "PVY310D"
 "PVY420D"
 "PVY520D"
 "SPR030D"
 "SPR130D"
 "SPR230D"
 "SPR330D"
 "SPR440D"
 "VST020D"

In [11]:
result = []

for i in 1:size(data,1)
    for j in 1:size(data,2)
        data[i,j] = round(data[i,j], digits=6)
        rdata[i,j] = round(rdata[i,j], digits=6)
    end
end

for col in column_names
    result = sum(data[!,col] .== rdata[!,col])

    if result .== 3900
        println("Column ", col,  " are equal at 6 decimals")
    else
        println("Column ", col,  " are not equal at 6 decimals")
    end

end

Column CountyCode are equal at 6 decimals
Column logghomr are equal at 6 decimals
Column logfssl are equal at 6 decimals
Column logrobr are equal at 6 decimals
Column logburg are equal at 6 decimals
Column burg_missing are equal at 6 decimals
Column robrate_missing are equal at 6 decimals
Column newblack are equal at 6 decimals
Column newfhh are equal at 6 decimals
Column newmove are equal at 6 decimals
Column newdens are equal at 6 decimals
Column newmal are equal at 6 decimals
Column AGE010D are equal at 6 decimals
Column AGE050D are equal at 6 decimals
Column AGE110D are equal at 6 decimals
Column AGE170D are equal at 6 decimals
Column AGE180D are equal at 6 decimals
Column AGE270D are equal at 6 decimals
Column AGE310D are equal at 6 decimals
Column AGE320D are equal at 6 decimals
Column AGE350D are equal at 6 decimals
Column AGE380D are equal at 6 decimals
Column AGE410D are equal at 6 decimals
Column AGE470D are equal at 6 decimals
Column AGE570D are equal at 6 decimals
Column AG

Now, we can construct the treatment variable, the outcome variable and the matrix $Z$ that includes the control variables.

In [12]:
# Treatment Variable
D = rdata[!, d]
println("Variable: <<D>> has ", size(D), " outcomes")

# Outcome Variable
Y = rdata[!, y]
println("Variable: <<Y>> has ", size(Y), " outcomes")

# Construct matrix Z
Z = rdata[!, vcat(X1, X2, census)]
println("Variable: <<Z>> has ", size(Z), " outcomes")

Variable: <<D>> has (3900,) outcomes
Variable: <<Y>> has (3900,) outcomes
Variable: <<Z>> has (3900, 195) outcomes


We have in total 195 control variables. The control variables $Z_{j,t}$ are from the U.S. Census Bureau and  contain demographic and economic characteristics of the counties such as  the age distribution, the income distribution, crime rates, federal spending, home ownership rates, house prices, educational attainment, voting paterns, employment statistics, and migration rates. 

In [13]:
clu = rdata[!, "CountyCode"] #for clustering the standard errors

3900-element Vector{Int64}:
  1073
  1073
  1073
  1073
  1073
  1073
  1073
  1073
  1073
  1073
  1073
  1073
  1073
     ⋮
 55133
 55133
 55133
 55133
 55133
 55133
 55133
 55133
 55133
 55133
 55133
 55133

In [14]:
data = DataFrame(hcat(clu, Y, D), [:CountyCode, :logghomr, :logfssl])
data = [data Z]

3900×198 DataFrame
  Row │ CountyCode  logghomr   logfssl    logrobr    logburg    burg_missing   ⋯
      │ Float64     Float64    Float64    Float64    Float64    Float64        ⋯
──────┼─────────────────────────────────────────────────────────────────────────
    1 │     1073.0  -0.134778   0.096127   0.150893  -0.124395      0.010461   ⋯
    2 │     1073.0  -0.239622   0.080809   0.040168  -0.134781      0.010461
    3 │     1073.0  -0.078677   0.05734   -0.017679  -0.167909      0.010461
    4 │     1073.0  -0.331465   0.081694  -0.009633  -0.22925       0.010461
    5 │     1073.0  -0.31664    0.025366  -0.026715  -0.176635      0.003248   ⋯
    6 │     1073.0   0.105132  -0.006777  -0.151487  -0.189069      0.010461
    7 │     1073.0  -0.03734    0.077306  -0.166729  -0.117739      0.010461
    8 │     1073.0  -0.052061  -0.108433  -0.099645  -0.083309      0.00449
    9 │     1073.0   0.054701  -0.034099   0.151557   0.319282     -0.044835   ⋯
   10 │     1073.0   0.122094  -0.082429   0.047603  -0.014473     -0.002332
   11 │     1073.0   0.110314  -0.024344   0.008143  -0.034969      0.010461
  ⋮   │     ⋮           ⋮          ⋮          ⋮          ⋮           ⋮         ⋱
 3891 │    55133.0  -0.54942   -0.006815   0.078647   0.13458      -0.020588
 3892 │    55133.0  -0.660408   0.135748   0.05618    0.134567     -0.021935   ⋯
 3893 │    55133.0  -0.676525  -0.022392   0.209323   0.390146     -0.06076
 3894 │    55133.0  -0.697525   0.040516  -0.085853   0.15756      -0.021679
 3895 │    55133.0  -0.03841   -0.058838  -0.09832    0.126737     -0.023104
 3896 │    55133.0   0.028096   0.040789   0.018225   0.107714     -0.030275   ⋯
 3897 │    55133.0   0.080145   0.208251   0.251402   0.151313     -0.030861
 3898 │    55133.0   0.497987   0.177444   0.179985   0.116497     -0.023712
 3899 │    55133.0   0.568542  -0.097127  -1.3754    -3.02228       0.530429
 3900 │    55133.0   0.609022   0.274062   0.336788   0.171876     -0.036207   ⋯
                                               192 columns and 3879 rows omitted

In [15]:
# CSV.write("../data/gun_clean2.csv", data)

## The effect of gun ownership

### OLS

After preprocessing the data, we first look at simple regression of $Y_{j,t}$ on $D_{j,t-1}$ without controls as a baseline model.

In [16]:
# # Run this line to avoid all the lines of code above
# data = CSV.File("../data/gun_clean2.csv") |> DataFrame

In [17]:
using FixedEffectModels

In [18]:
#baseline_formula <- as.formula(paste(y, "~", d ))
#baseline.ols <- lm(baseline_formula,data=rdata)

fm_1 = @formula(logghomr ~ 0 + logfssl + fe(CountyCode))
baseline_ols = reg(data, fm_1, Vcov.cluster(:CountyCode))

                        Fixed Effect Model                        
Number of obs:              3900  Degrees of freedom:            2
R2:                        0.006  R2 Adjusted:               0.006
F-Stat:                  18.9732  p-value:                   0.000
R2 within:                 0.006  Iterations:                    1
logghomr | Estimate Std.Error t value Pr(>|t|) Lower 95% Upper 95%
------------------------------------------------------------------
logfssl  | 0.282304 0.0648108 4.35582    0.000  0.155238   0.40937


In [19]:
println("2.5% : ", coeftable(baseline_ols).cols[5])
println("97.5% : " , coeftable(baseline_ols).cols[6])
println("Estimate: ", coeftable(baseline_ols).cols[1])
println("Cluster s.e. : " , r2(baseline_ols))
println("T-value : ", coeftable(baseline_ols).cols[3])
println("Pr(>|t|) : " , coeftable(baseline_ols).cols[4])

2.5% : [0.15523789539597105]
97.5% : [0.4093704387686366]
Estimate: [0.2823041670823038]
Cluster s.e. : 0.006193251272214595
T-value : [4.355820406592104]
Pr(>|t|) : [1.3597928318408193e-5]


The point estimate is $0.282$ with the confidence interval ranging from 0.155 to 0.41. This
suggests that increases in gun ownership rates are related to gun homicide rates - if gun ownership increases by 1% relative
to a trend then the predicted gun homicide rate goes up by 0.28%, without controlling for counties' characteristics.

Since our goal is to estimate the effect of gun ownership after controlling for a rich set county characteristics we next include the controls. First, we estimate the model by ols and then by an array of the modern regression methods using the double machine learning approach.

In [20]:
# define the variables
y = "logghomr"

data_columns = names(data)
no_relev_col = ["CountyCode", "logghomr"]

# This gives us: new_list = ['carrot' , 'lemon']
z = data[!,Not(no_relev_col)]

3900×196 DataFrame
  Row │ logfssl    logrobr    logburg    burg_missing  robrate_missing  newbla ⋯
      │ Float64    Float64    Float64    Float64       Float64          Float6 ⋯
──────┼─────────────────────────────────────────────────────────────────────────
    1 │  0.096127   0.150893  -0.124395      0.010461        -0.021229   0.030 ⋯
    2 │  0.080809   0.040168  -0.134781      0.010461        -0.019418   0.030
    3 │  0.05734   -0.017679  -0.167909      0.010461        -0.022037   0.030
    4 │  0.081694  -0.009633  -0.22925       0.010461        -0.019418   0.030
    5 │  0.025366  -0.026715  -0.176635      0.003248        -0.020804   0.030 ⋯
    6 │ -0.006777  -0.151487  -0.189069      0.010461         0.016953   0.030
    7 │  0.077306  -0.166729  -0.117739      0.010461         0.02455    0.030
    8 │ -0.108433  -0.099645  -0.083309      0.00449          0.021457   0.030
    9 │ -0.034099   0.151557   0.319282     -0.044835        -0.036663   0.030 ⋯
   10 │ -0.082429   0.047603  -0.014473     -0.002332         0.007654   0.030
   11 │ -0.024344   0.008143  -0.034969      0.010461         0.010167  -0.030
  ⋮   │     ⋮          ⋮          ⋮           ⋮               ⋮             ⋮  ⋱
 3891 │ -0.006815   0.078647   0.13458      -0.020588        -0.020788   0.053
 3892 │  0.135748   0.05618    0.134567     -0.021935        -0.016185   0.053 ⋯
 3893 │ -0.022392   0.209323   0.390146     -0.06076         -0.011997   0.053
 3894 │  0.040516  -0.085853   0.15756      -0.021679        -0.016735   0.053
 3895 │ -0.058838  -0.09832    0.126737     -0.023104        -0.017127   0.053
 3896 │  0.040789   0.018225   0.107714     -0.030275        -0.027317   0.053 ⋯
 3897 │  0.208251   0.251402   0.151313     -0.030861        -0.041741   0.053
 3898 │  0.177444   0.179985   0.116497     -0.023712        -0.023737   0.053
 3899 │ -0.097127  -1.3754    -3.02228       0.530429         0.530404   0.053
 3900 │  0.274062   0.336788   0.171876     -0.036207        -0.030526   0.053 ⋯
                                               191 columns and 3879 rows omitted

In [26]:
names(FixedEffectModels)

63-element Vector{Symbol}:
 Symbol("@formula")
 :AbstractContrasts
 :AbstractTerm
 :CategoricalTerm
 :ConstantTerm
 :ContinuousTerm
 :ContrastsCoding
 :DummyCoding
 :EffectsCoding
 :FixedEffectModel
 :FixedEffectModels
 :FormulaTerm
 :FunctionTerm
 ⋮
 :reg
 :residuals
 :response
 :responsename
 :rss
 :schema
 :setcontrasts!
 :stderror
 :term
 :terms
 :vcov
 :width

In [27]:
# control_formula = term(:logghomr) ~ sum(term.(names(z)[1:end])) + fe(:CountyCode)
control_formula = term(:logghomr) ~ sum(term.(names(z)[1:end]))

FormulaTerm
Response:
  logghomr(unknown)
Predictors:
  logfssl(unknown)
  logrobr(unknown)
  logburg(unknown)
  burg_missing(unknown)
  robrate_missing(unknown)
  newblack(unknown)
  newfhh(unknown)
  newmove(unknown)
  newdens(unknown)
  newmal(unknown)
  AGE010D(unknown)
  AGE050D(unknown)
  AGE110D(unknown)
  AGE170D(unknown)
  AGE180D(unknown)
  AGE270D(unknown)
  AGE310D(unknown)
  AGE320D(unknown)
  AGE350D(unknown)
  AGE380D(unknown)
  AGE410D(unknown)
  AGE470D(unknown)
  AGE570D(unknown)
  AGE640D(unknown)
  AGE670D(unknown)
  AGE760D(unknown)
  BNK010D(unknown)
  BNK050D(unknown)
  BPS030D(unknown)
  BPS130D(unknown)
  BPS230D(unknown)
  BPS020D(unknown)
  BPS120D(unknown)
  BPS220D(unknown)
  BPS820D(unknown)
  BZA010D(unknown)
  BZA110D(unknown)
  BZA210D(unknown)
  EDU100D(unknown)
  EDU200D(unknown)
  EDU600D(unknown)
  EDU610D(unknown)
  EDU620D(unknown)
  EDU630D(unknown)
  EDU635D(unknown)
  EDU640D(unknown)
  EDU650D(unknown)
  EDU680D(unknown)
  EDU685D(unknown)
  E

In [28]:
control_ols = reg(data, control_formula, Vcov.cluster(:CountyCode))

┌ Warning: estimated covariance matrix of moment conditions not of full rank.
│                  model tests should be interpreted with caution.
└ @ Vcov C:\Users\sandr\.julia\packages\Vcov\8Fkqk\src\utils.jl:5


                                   Linear Model                                   
Number of obs:                      3900   Degrees of freedom:                  181
R2:                                0.203   R2 Adjusted:                       0.164
F-Stat:                       2.71826e18   p-value:                           0.000
logghomr        |    Estimate   Std.Error      t value Pr(>|t|) Lower 95% Upper 95%
-----------------------------------------------------------------------------------
logfssl         |    0.190671 0.000214096      890.584    0.000  0.190251   0.19109
logrobr         |     0.18903     9.74947    0.0193888    0.985  -18.9258   19.3039
logburg         |    0.219294     4.47891    0.0489614    0.961  -8.56206   9.00065
burg_missing    |      1.5298     7.52835     0.203205    0.839  -13.2303   16.2899
robrate_missing |     1.13297     7.26053     0.156044    0.876  -13.1021    15.368
newblack        |    -4.34682     6.62062    -0.656558    0.512  -17.3272   8

In [29]:
println("For <<logfssl>> variable: ")
println("2.5% : ", coeftable(control_ols).cols[5][1])
println("97.5% : " , coeftable(control_ols).cols[6][1])
println("Estimate: ", coeftable(control_ols).cols[1][1])
println("Cluster s.e. : " , r2(control_ols))
println("T-value : ", coeftable(control_ols).cols[3][1])
println("Pr(>|t|) : " , coeftable(control_ols).cols[4][1])

For <<logfssl>> variable: 
2.5% : 0.19025077455652542
97.5% : 0.19109028864460129
Estimate: 0.19067053160056335
Cluster s.e. : 0.20296178713656843
T-value : 890.5843124451466
Pr(>|t|) : 0.0


After controlling for a rich set of characteristics, the point estimate of gun ownership reduces to $0.19$.

# DML algorithm

Here we perform inference of the predictive coefficient $\beta$ in our partially linear statistical model, 

$$
Y = D\beta + g(Z) + \epsilon, \quad E (\epsilon | D, Z) = 0,
$$

using the **double machine learning** approach. 

For $\tilde Y = Y- E(Y|Z)$ and $\tilde D= D- E(D|Z)$, we can write
$$
\tilde Y = \alpha \tilde D + \epsilon, \quad E (\epsilon |\tilde D) =0.
$$

Using cross-fitting, we employ modern regression methods
to build estimators $\hat \ell(Z)$ and $\hat m(Z)$ of $\ell(Z):=E(Y|Z)$ and $m(Z):=E(D|Z)$ to obtain the estimates of the residualized quantities:

$$
\tilde Y_i = Y_i  - \hat \ell (Z_i),   \quad \tilde D_i = D_i - \hat m(Z_i), \quad \text{ for each } i = 1,\dots,n.
$$

Finally, using ordinary least squares of $\tilde Y_i$ on $\tilde D_i$, we obtain the 
estimate of $\beta$.

The following algorithm comsumes $Y, D, Z$, and a machine learning method for learning the residuals $\tilde Y$ and $\tilde D$, where the residuals are obtained by cross-validation (cross-fitting). Then, it prints the estimated coefficient $\beta$ and the corresponding standard error from the final OLS regression.

In [994]:
# import Pkg; Pkg.add("MLDataUtils")
# import Pkg; Pkg.add("MLBase")

In [ ]:
using  MLDataUtils, MLBase

In [1016]:
function DML2_for_PLM(z, d, y, dreg, yreg, nfold, clu) 
    
    # Num observations
    nobser = size(z,1)
    
    # Define folds indices 
    list_1 = repeat(collect(1:1:nfold), nobser)
    samplea = randperm(nobser)
    foldid = [list_1[i] for i in samplea]

    println("Folds: " )
    
    # Create split function(similar to R)
    function split(f, n)
        a = []
        for i in 1:n
            push!(a ,findall(x -> x == i, f))
        end
        a
    end 

    # Split observation indices into folds 
    I = split(foldid, nfold)
     
    # Create array to save errors 
    ytil = ones(nobser)
    dtil = ones(nobser)

    # loop to save results
    for i in 1:nfold 
        dfit = dreg(Z, D, rdata_1[Not(I[i]),:])
        yfit = yreg(Z, Y, rdata_1[Not(I[i]),:])
        dhat = predict(dfit, Z[(I[i]),:])
        yhat = predict(yfit, Z[(I[i]),:])
        dtil[(I[i]),:] .= (D[(I[i]),:] .- dhat)
        ytil[(I[i]),:] .= (Y[(I[i]),:] .- yhat)
        println(i)
    end
    
    data = DataFrame(ytil = ytil, dtil = dtil, clu = clu)
    rfit = reg(data, @formula(ytil ~ dtil + fe(clu)))
    coef_est = coef(rfit)[1]
    se = FixedEffectModels.coeftable(rfit).cols[2]
    
    println(" coef (se) = ", coef_est ,"(",se,")")
    
end

DML2_for_PLM (generic function with 1 method)

Now, we apply the Double Machine Learning (DML) approach with different machine learning methods. First, we load the relevant libraries.

Let us, construct the input matrices.

In [742]:
# Create main variables
Y = rdata[!, "logghomr"]
Y = DataFrame([Y], [:Y])

D = rdata[!, "logfssl"]
D = DataFrame([D], [:D])

clu = rdata[!, "CountyCode"]
clu = DataFrame([clu], [:CountyCode])

Z = select(rdata, Not(["logghomr", "logfssl", "CountyCode"]))

rdata_1 = [Y D Z]

first(DataFrame([Y D clu]), 6)
# first(DataFrame(logghomr = y,logfssl = d,CountyCode = clu ),6)

6×3 DataFrame
 Row │ Y          D          CountyCode 
     │ Float64    Float64    Int64      
─────┼──────────────────────────────────
   1 │ -0.134778   0.096127        1073
   2 │ -0.239622   0.080809        1073
   3 │ -0.078677   0.05734         1073
   4 │ -0.331465   0.081694        1073
   5 │ -0.31664    0.025366        1073
   6 │  0.105132  -0.006777        1073

In the following, we apply the DML approach with the differnt versions of lasso.


## Lasso

In [854]:
using Lasso, GLMNet

In [1017]:
# DML with Lasso:

Random.seed!(123)

function dreg(Z, D, rdata_1)
    fit(LassoModel, term(:D) ~  sum(term.(names(Z))), rdata_1; α = 0.8, standardize = false)
end

############################

function yreg(Z, Y, rdata_1)
    fit(LassoModel, term(:Y) ~  sum(term.(names(Z))), rdata_1; α = 0.8, standardize = false)
end

DML2_for_PLM(z, d, y, dreg, yreg, 10, clu)

Folds: 
1
2
3
4
5
6
7
8
9
10
 coef (se) = 0.2114887327350614([0.055559216760055416])


In [ ]:
# # DML with Post-Lasso:

# Random.seed!(123)

# function dreg(Z, D, rdata_1)
#     a = fit(LassoModel, term(:D) ~  sum(term.(names(Z))), rdata_1; α = 0.8, standardize = false)
#     new_d = DataFrame([coef(a)], [:new_d])
# end

# ############################

# function yreg(Z, Y, rdata_1)
#     fit(LassoModel, term(:Y) ~  sum(term.(names(Z))), rdata_1; α = 0.8, standardize = false)
# end

# DML2_for_PLM(z, d, y, dreg, yreg, 10, clu)

In [856]:
# DML with cross-validated Lasso:
names(GLMNet)

12-element Vector{Symbol}:
 :CompressedPredictorMatrix
 :CoxPH
 :GLMNet
 :GLMNetCrossValidation
 :GLMNetPath
 :coef
 :glmnet
 :glmnet!
 :glmnetcv
 :lambdamin
 :nactive
 :predict

In [1048]:
nobser = 10
nfold = 5

5

In [1047]:
foldid[1,:]

1-element Vector{Any}:
 [1, 2, 3, 4, 5, 8, 9, 10]

In [1045]:
foldid = collect(Kfold(nobser, nfold))

5-element Vector{Any}:
 [1, 2, 3, 4, 5, 8, 9, 10]
 [1, 3, 4, 6, 7, 8, 9, 10]
 [2, 3, 5, 6, 7, 8, 9, 10]
 [1, 2, 4, 5, 6, 7, 9, 10]
 [1, 2, 3, 4, 5, 6, 7, 8]

In [ ]:
nobser = 10
nfold = 5

In [1049]:


    # Define folds indices 
    list_1 = repeat(collect(1:1:nfold), nobser)
    samplea = randperm(nobser)
    foldid = [list_1[i] for i in samplea]

    println("Folds: " )
    
    # Create split function(similar to R)
    function split(f, n)
        a = []
        for i in 1:n
            push!(a ,findall(x -> x == i, f))
        end
        a
    end 

    # Split observation indices into folds 
    I = split(foldid, nfold)
     

Folds: 


5-element Vector{Any}:
 [2, 10]
 [5, 9]
 [4, 6]
 [7, 8]
 [1, 3]

In [ ]:
    ytil = ones(nobser)
    dtil = ones(nobser)
    println("Folds: " )

In [ ]:
function DML2_lasso_cv(z, d, y, dreg, yreg, nfold, clu) 
    
    # Num observations
    nobser = size(z,1)
    
    # Define folds indices 
    list_1 = repeat(collect(1:1:nfold), nobser)
    samplea = randperm(nobser)
    foldid = [list_1[i] for i in samplea]

    println("Folds: " )
    
    # Create split function(similar to R)
    function split(f, n)
        a = []
        for i in 1:n
            push!(a ,findall(x -> x == i, f))
        end
        a
    end 

    # Split observation indices into folds 
    I = split(foldid, nfold)
     
    # Create array to save errors 
    ytil = ones(nobser)
    dtil = ones(nobser)

    # loop to save results
    for i in 1:nfold 
        dfit = dreg(Z, D, rdata_1[Not(I[i]),:])
        yfit = yreg(Z, Y, rdata_1[Not(I[i]),:])
        dhat = predict(dfit, Z[(I[i]),:])
        yhat = predict(yfit, Z[(I[i]),:])
        dtil[(I[i]),:] .= (D[(I[i]),:] .- dhat)
        ytil[(I[i]),:] .= (Y[(I[i]),:] .- yhat)
        println(i)
    end
    
    data = DataFrame(ytil = ytil, dtil = dtil, clu = clu)
    rfit = reg(data, @formula(ytil ~ dtil + fe(clu)))
    coef_est = coef(rfit)[1]
    se = FixedEffectModels.coeftable(rfit).cols[2]
    
    println(" coef (se) = ", coef_est ,"(",se,")")
    
end

In [942]:
function dreg(z,d)
    glmnetcv(z,d, alpha = 1)    
end
function yreg(z,y)
    glmnetcv(z,y, alpha = 1)    
end

yreg (generic function with 2 methods)

In [941]:
function DML2_lasso_cv(z, d, y, dreg, yreg, nfold, clu)
    nobser = size(z,1) #number of observations
    foldid = collect(Kfold(size(z)[1], nfold))
    ytil = ones(nobser)
    dtil = ones(nobser)
    println("Folds: " )
    for i in 1:nfold
        dfit = dreg(z[foldid[i],:], d[foldid[i]])
        yfit = yreg(z[foldid[i],:], y[foldid[i]])
        dhat = GLMNet.predict(dfit, z[Not(foldid[i]),:])
        yhat = GLMNet.predict(yfit, z[Not(foldid[i]),:])
        dtil[Not(foldid[i])]   = (d[Not(foldid[i])] - dhat)
        ytil[Not(foldid[i])]   = (y[Not(foldid[i])] - yhat)
        println(i)
    end
    
    data = DataFrame(ytil = ytil, dtil = dtil, clu=clu)
    rfit = reg(data, @formula(ytil ~ dtil +fe(clu)))
    coef_est = coef(rfit)[1]
    se = FixedEffectModels.coeftable(rfit).cols[2]

    println(" coef (se) = ", coef_est ,"(",se,")")
    
end

DML2_lasso_cv (generic function with 1 method)

In [944]:

DML2_lasso_cv(z,d,y,dreg,yreg,5,clu)

Folds: 
1
2
3
4
5
 coef (se) = 0.20397859201961707([0.055608594106786605])


In [947]:
z = Matrix(Z)
d = D[!,1]
y = Y[!,1]

3900-element Vector{Float64}:
 -0.134778
 -0.239622
 -0.078677
 -0.331465
 -0.31664
  0.105132
 -0.03734
 -0.052061
  0.054701
  0.122094
  0.110314
  0.256268
  0.079057
  ⋮
  0.214936
 -0.490341
 -0.54942
 -0.660408
 -0.676525
 -0.697525
 -0.03841
  0.028096
  0.080145
  0.497987
  0.568542
  0.609022

In [948]:
 a = glmnetcv(z, d, alpha = 1)

Least Squares GLMNet Cross Validation
100 models for 195 predictors in 10 folds
Best λ 0.002 (mean loss 0.013, std 0.001)

In [969]:
row.index(rdata_1[I[1],:])

LoadError: UndefVarError: row not defined

In [988]:
a = glmnetcv(z, d, alpha = 1)

Least Squares GLMNet Cross Validation
100 models for 195 predictors in 10 folds
Best λ 0.002 (mean loss 0.013, std 0.001)

In [1031]:
z[[1,2],:]

2×195 Matrix{Float64}:
 0.150893  -0.124395  0.010461  -0.021229  …  0.026761  -0.040704  0.077142
 0.040168  -0.134781  0.010461  -0.019418     0.026761  -0.040704  0.077142

In [1032]:
Matrix(z[[1,2],:])

2×195 Matrix{Float64}:
 0.150893  -0.124395  0.010461  -0.021229  …  0.026761  -0.040704  0.077142
 0.040168  -0.134781  0.010461  -0.019418     0.026761  -0.040704  0.077142

In [1036]:
dreg(z[[1,2],:], d[[1,2],:])

LoadError: DivideError: integer division error

In [1034]:
a = glmnetcv(z[[1,2],:], d[[1,2],], alpha = 1)

LoadError: DivideError: integer division error

In [1035]:
function dreg(z,d,subset)
    a = glmnetcv(z, d, alpha = 1)
    return GLMNet.coef(a)
end

dreg (generic function with 2 methods)

In [1021]:
dfit = dreg(Z, D)

LoadError: MethodError: no method matching glmnetcv(::DataFrame, ::DataFrame; alpha=1)
Closest candidates are:
  glmnetcv(::AbstractMatrix{T} where T, ::Any; kw...) at C:\Users\sandr\.julia\packages\GLMNet\C8WKF\src\Multinomial.jl:205

In [1020]:
dfit = dreg(Z, D, rdata_1[Not(I[i]),:])

LoadError: UndefVarError: i not defined

In [ ]:
yfit = yreg(Z, Y, rdata_1[Not(I[i]),:])

In [ ]:
dhat = predict(dfit, Z[(I[i]),:])
yhat = predict(yfit, Z[(I[i]),:])
dtil[(I[i]),:] .= (D[(I[i]),:] .- dhat)
ytil[(I[i]),:] .= (Y[(I[i]),:] .- yhat)
println(i)

In [976]:
dfit = dreg(z,d,I[2])

2-element Vector{Float64}:
 -0.05695885579177813
  0.0

In [993]:
bb = GLMNet.coef(a)

195-element Vector{Float64}:
  0.008180988198113932
  0.0
  0.0
  0.0
  0.0
  0.0
  0.0
 -0.08586749450375761
 -0.03297444162310575
  0.0
 -0.2743304944607936
  0.0
  0.0
  ⋮
  0.0
  0.0
  0.0
  0.0
  0.0
  0.0
  0.0
  0.0
  0.0
  0.0
  0.0
  0.04747464058243396

In [982]:
names(GLMNet)

12-element Vector{Symbol}:
 :CompressedPredictorMatrix
 :CoxPH
 :GLMNet
 :GLMNetCrossValidation
 :GLMNetPath
 :coef
 :glmnet
 :glmnet!
 :glmnetcv
 :lambdamin
 :nactive
 :predict

In [991]:
z[Not(I[1]),:]

3897×195 Matrix{Float64}:
  0.150893  -0.124395   0.010461  …   0.026761  -0.040704   0.077142
  0.040168  -0.134781   0.010461      0.026761  -0.040704   0.077142
 -0.017679  -0.167909   0.010461      0.026761  -0.040704   0.077142
 -0.009633  -0.22925    0.010461      0.026761  -0.040704   0.077142
 -0.026715  -0.176635   0.003248      0.026761  -0.040704   0.077142
 -0.151487  -0.189069   0.010461  …   0.026761  -0.040704  -0.00456
 -0.166729  -0.117739   0.010461      0.026761  -0.040704  -0.00456
 -0.099645  -0.083309   0.00449       0.026761  -0.040704  -0.00456
  0.047603  -0.014473  -0.002332      0.026761  -0.040704  -0.00456
  0.008143  -0.034969   0.010461      0.026761  -0.040704  -0.058292
 -0.081185   0.044147   0.008486  …  -0.009951  -0.006583  -0.055527
 -0.153504   0.393015  -0.048444     -0.021438   0.071247  -0.062605
  0.024709   0.053162   0.006772     -0.036801   0.061972  -0.05856
  ⋮                               ⋱                        
 -0.118829   0.299746 

In [992]:
a

Least Squares GLMNet Cross Validation
100 models for 195 predictors in 10 folds
Best λ 0.002 (mean loss 0.013, std 0.001)

In [990]:
GLMNet.predict(a,z[Not(I[1]),:],alpha=1)

LoadError: MethodError: no method matching predict(::GLMNetPath{Normal{Float64}, Vector{Float64}, CompressedPredictorMatrix}, ::Matrix{Float64}, ::Int64; outtype=:link, alpha=1)
Closest candidates are:
  predict(::GLMNetPath, ::AbstractMatrix{T} where T, ::Union{Int64, AbstractVector{Int64}}; outtype, offsets) at C:\Users\sandr\.julia\packages\GLMNet\C8WKF\src\GLMNet.jl:105 got unsupported keyword argument "alpha"
  predict(::GLMNetPath{var"#s85", A, B} where {var"#s85"<:Multinomial, A<:AbstractArray, B<:AbstractArray}, ::AbstractMatrix{T} where T, ::Union{Int64, AbstractVector{Int64}}; outtype, offsets) at C:\Users\sandr\.julia\packages\GLMNet\C8WKF\src\Multinomial.jl:12 got unsupported keyword argument "alpha"
  predict(::GLMNetPath{var"#s88", A, B} where {var"#s88"<:CoxPH, A<:AbstractArray, B<:AbstractArray}, ::AbstractMatrix{T} where T, ::Union{Int64, AbstractVector{Int64}}; outtype, offsets) at C:\Users\sandr\.julia\packages\GLMNet\C8WKF\src\CoxNet.jl:30 got unsupported keyword argument "alpha"
  ...

In [978]:
        dhat = GLMNet.predict(z, dfit, I[2] )

LoadError: MethodError: no method matching predict(::Matrix{Float64}, ::Vector{Float64}, ::Vector{Int64})
Closest candidates are:
  predict(::GLMNetPath{var"#s85", A, B} where {var"#s85"<:Multinomial, A<:AbstractArray, B<:AbstractArray}, ::AbstractMatrix{T} where T, ::Union{Int64, AbstractVector{Int64}}; outtype, offsets) at C:\Users\sandr\.julia\packages\GLMNet\C8WKF\src\Multinomial.jl:12
  predict(::GLMNetPath{var"#s88", A, B} where {var"#s88"<:CoxPH, A<:AbstractArray, B<:AbstractArray}, ::AbstractMatrix{T} where T, ::Union{Int64, AbstractVector{Int64}}; outtype, offsets) at C:\Users\sandr\.julia\packages\GLMNet\C8WKF\src\CoxNet.jl:30
  predict(::GLMNetPath, ::AbstractMatrix{T} where T, ::Union{Int64, AbstractVector{Int64}}; outtype, offsets) at C:\Users\sandr\.julia\packages\GLMNet\C8WKF\src\GLMNet.jl:105

In [965]:
a = dreg(z[I[4],:],d[I[4]],subset)

LoadError: DivideError: integer division error

In [945]:
    GLMNet.coef(a)[I[1]]

3-element Vector{Float64}:
 -0.031184368871806183
  0.0
  0.0

In [946]:
    GLMNet.predict(a)[I[1]]

LoadError: MethodError: no method matching predict(::GLMNetCrossValidation)
Closest candidates are:
  predict(::GLMNetCrossValidation, ::AbstractMatrix{T} where T; outtype, kw...) at C:\Users\sandr\.julia\packages\GLMNet\C8WKF\src\CoxNet.jl:167
  predict(::GLMNetPath{var"#s86", A, B} where {var"#s86"<:Multinomial, A<:AbstractArray, B<:AbstractArray}, ::AbstractMatrix{T} where T) at C:\Users\sandr\.julia\packages\GLMNet\C8WKF\src\Multinomial.jl:12
  predict(::GLMNetPath{var"#s85", A, B} where {var"#s85"<:Multinomial, A<:AbstractArray, B<:AbstractArray}, ::AbstractMatrix{T} where T, ::Union{Int64, AbstractVector{Int64}}; outtype, offsets) at C:\Users\sandr\.julia\packages\GLMNet\C8WKF\src\Multinomial.jl:12
  ...

In [917]:
zz .= Matrix(Z[I[3],:])

2×195 Matrix{Float64}:
 -0.166729  -0.117739  0.010461  0.02455   …  0.026761  -0.040704  -0.00456
  0.008143  -0.034969  0.010461  0.010167     0.026761  -0.040704  -0.058292

In [915]:
dd.=D[I[3],1]

2-element Vector{Float64}:
  0.077306
 -0.024344

In [918]:
a = glmnetcv(zz, dd)

LoadError: DivideError: integer division error

In [875]:
GLMNetCrossValidation(a)

LoadError: MethodError: no method matching GLMNetCrossValidation(::GLMNetCrossValidation)
Closest candidates are:
  GLMNetCrossValidation(::Any, ::Int64, ::Vector{Float64}, ::Vector{Float64}, ::Vector{Float64}, ::Vector{Vector{Float64}}) at C:\Users\sandr\.julia\packages\GLMNet\C8WKF\src\GLMNet.jl:499
  GLMNetCrossValidation(::Any, ::Any, ::Any, ::Any, ::Any, ::Any) at C:\Users\sandr\.julia\packages\GLMNet\C8WKF\src\GLMNet.jl:499

In [ ]:
function dreg(z,d, rdata_1)
    glmnetcv(Z,D, rdata_1, alpha = 1)    
end
function yreg(z,y)
    glmnetcv(z,y, rdata_1, alpha = 1)    
end

In [94]:
#DML with cross-validated Lasso:
dreg <- function(z,d){ cv.glmnet(z,d,family="gaussian", alpha=1) } #ML method = lasso from glmnet 
yreg <- function(z,y){ cv.glmnet(z,y,family="gaussian", alpha=1) }  #ML method = lasso from glmnet 
DML2.lasso.cv = DML2.for.PLM(z, d, y, dreg, yreg, nfold=5, clu)

dreg <- function(z,d){ cv.glmnet(z,d,family="gaussian", alpha=0.5) } #ML method = elastic net from glmnet 
yreg <- function(z,y){ cv.glmnet(z,y,family="gaussian", alpha=0.5) }  #ML method = elastic net from glmnet 
DML2.elnet = DML2.for.PLM(z, d, y, dreg, yreg, nfold=5, clu)

dreg <- function(z,d){ cv.glmnet(z,d,family="gaussian", alpha=0) } #ML method = ridge from glmnet 
yreg <- function(z,y){ cv.glmnet(z,y,family="gaussian", alpha=0) }  #ML method = ridge from glmnet 
DML2.ridge = DML2.for.PLM(z, d, y, dreg, yreg, nfold=5, clu)

fold: 1  2  3  4  5  
coef (se) = 0.194926 (0.0569378)
fold: 1  2  3  4  5  
coef (se) = 0.208474 (0.0600804)
fold: 1  2  3  4  5  
coef (se) = 0.200234 (0.0598422)


Here we also compute DML with OLS used as the ML method

In [95]:
dreg <- function(z,d){  glmnet(z,d,family="gaussian", lambda=0) } #ML method = ols from glmnet 
yreg <- function(z,y){  glmnet(z,y,family="gaussian", lambda=0) }  #ML method = ols from glmnet 
DML2.ols = DML2.for.PLM(z, d, y, dreg, yreg, nfold=10, clu)

fold: 1  2  3  4  5  6  7  8  9  10  
coef (se) = 0.203079 (0.051136)


Next, we also apply Random Forest for comparison purposes.

### Random Forest


In [66]:
#DML with Random Forest:
dreg <- function(z,d){ randomForest(z, d) } #ML method=Forest 
yreg <- function(z,y){ randomForest(z, y) } #ML method=Forest
set.seed(1)
DML2.RF = DML2.for.PLM(z, d, y, dreg, yreg, nfold=2, clu) # set to 2 due to computation time

fold: 1  2  
coef (se) = 0.153017 (0.0605311)


In [91]:
if (!is.null(d) && !is.factor(d))
             max(floor(ncol(z)/3), 1) else floor(sqrt(ncol(z)))

[1] 65

In [92]:
if (!is.null(d) && !is.factor(d)) 5 else 1

[1] 5

We conclude that the gun ownership rates are related to gun homicide rates - if gun ownership increases by 1% relative
to a trend then the predicted gun homicide rate goes up by about 0.20% controlling for counties' characteristics.

Finally, let's see which method is actually better. We compute RMSE for predicting D and Y, and see which
of the methods works better.


In [96]:
mods<- list(DML2.ols, DML2.lasso, DML2.post, DML2.lasso.cv, DML2.ridge, DML2.elnet, DML2.RF)

RMSE.mdl<- function(mdl) {
RMSEY <- sqrt(mean(mdl$ytil)^2) 
RMSED <- sqrt(mean(mdl$dtil)^2) 
return( list(RMSEY=RMSEY, RMSED=RMSED))
}

#RMSE.mdl(DML2.lasso)

#DML2.lasso$ytil

Res<- lapply(mods, RMSE.mdl)


prRes.Y<- c( Res[[1]]$RMSEY,Res[[2]]$RMSEY, Res[[3]]$RMSEY, Res[[4]]$RMSEY, Res[[5]]$RMSEY,  Res[[6]]$RMSEY, Res[[7]]$RMSEY)
prRes.D<- c( Res[[1]]$RMSED,Res[[2]]$RMSED, Res[[3]]$RMSED, Res[[4]]$RMSED, Res[[5]]$RMSED, Res[[6]]$RMSED, Res[[7]]$RMSED)

prRes<- rbind(prRes.Y, prRes.D); 
rownames(prRes)<- c("RMSE D", "RMSE Y");
colnames(prRes)<- c("OLS", "Lasso", "Post-Lasso", "CV Lasso", "CV Ridge", "CV Elnet", "RF")
print(prRes,digit=6)

               OLS       Lasso  Post-Lasso    CV Lasso    CV Ridge    CV Elnet
RMSE D 0.000407561 3.25471e-05 1.32656e-04 0.000376929 7.24337e-04 9.66559e-04
RMSE Y 0.000134575 3.35791e-05 6.89649e-05 0.000044933 8.41741e-05 5.80479e-19
               RF
RMSE D 0.01086246
RMSE Y 0.00152755


It looks like the best method for predicting D is Lasso, and the best method for predicting Y is CV Ridge.


In [97]:
dreg <- function(z,d){ rlasso(z,d, post=T) } #ML method= lasso from hdm 
yreg <- function(z,y){ cv.glmnet(z,y,family="gaussian", alpha=0) }  #ML method = ridge from glmnet 
DML2.best= DML2.for.PLM(z, d, y, dreg, yreg, nfold=10, clu)

fold: 1  2  3  4  5  6  7  8  9  10  
coef (se) = 0.222066 (0.0565614)


Let's organize the results in a table.

In [84]:
est_baseline

Estimate Cluster s.e.      t value     Pr(>|t|) 
2.823045e-01 6.481080e-02 4.355825e+00 1.359767e-05

In [18]:
library(xtable)

table <- matrix(0,9,2)
table[1,1] <- as.numeric(est_baseline[1])
table[2,1] <- as.numeric(est_ols[1])
table[3,1]   <- as.numeric(DML2.lasso$coef.est)
table[4,1]   <- as.numeric(DML2.post$coef.est)
table[5,1]  <-as.numeric(DML2.lasso.cv$coef.est)
table[6,1] <-as.numeric(DML2.elnet$coef.est)
table[7,1] <-as.numeric(DML2.ridge$coef.est)
table[8,1] <-as.numeric(DML2.RF$coef.est)
table[9,1] <-as.numeric(DML2.best$coef.est)
table[1,2] <- as.numeric(est_baseline[2])
table[2,2] <- as.numeric(est_ols[2])
table[3,2]   <- as.numeric(DML2.lasso$se)
table[4,2]   <- as.numeric(DML2.post$se)
table[5,2]  <-as.numeric(DML2.lasso.cv$se)
table[6,2] <-as.numeric(DML2.elnet$se)
table[7,2] <-as.numeric(DML2.ridge$se)
table[8,2] <-as.numeric(DML2.RF$se)
table[9,2] <-as.numeric(DML2.best$se)




################################# Print Results #################################

colnames(table) <- c("Estimate","Standard Error")
rownames(table) <- c("Baseline OLS", "Least Squares with controls", "Lasso", "Post-Lasso", "CV Lasso","CV Elnet", "CV Ridge", "Random Forest", 
                     "Best")

table

Estimate  Standard Error
Baseline OLS                0.2823045 0.06481080    
Least Squares with controls 0.1906447 0.05244756    
Lasso                       0.2228074 0.05702673    
Post-Lasso                  0.2269338 0.05619181    
CV Lasso                    0.2004742 0.05764115    
CV Elnet                    0.2061170 0.05746222    
CV Ridge                    0.2013789 0.05790663    
Random Forest               0.1921739 0.05814101    
Best                        0.2190048 0.05721956

In [19]:
print(table, digit=3)


                            Estimate Standard Error
Baseline OLS                   0.282         0.0648
Least Squares with controls    0.191         0.0524
Lasso                          0.223         0.0570
Post-Lasso                     0.227         0.0562
CV Lasso                       0.200         0.0576
CV Elnet                       0.206         0.0575
CV Ridge                       0.201         0.0579
Random Forest                  0.192         0.0581
Best                           0.219         0.0572


In [20]:
tab<- xtable(table, digits=3)
print(tab, type="latex")

% latex table generated in R 3.6.3 by xtable 1.8-4 package
% Sat Feb 13 17:41:19 2021
\begin{table}[ht]
\centering
\begin{tabular}{rrr}
  \hline
 & Estimate & Standard Error \\ 
  \hline
Baseline OLS & 0.282 & 0.065 \\ 
  Least Squares with controls & 0.191 & 0.052 \\ 
  Lasso & 0.223 & 0.057 \\ 
  Post-Lasso & 0.227 & 0.056 \\ 
  CV Lasso & 0.200 & 0.058 \\ 
  CV Elnet & 0.206 & 0.057 \\ 
  CV Ridge & 0.201 & 0.058 \\ 
  Random Forest & 0.192 & 0.058 \\ 
  Best & 0.219 & 0.057 \\ 
   \hline
\end{tabular}
\end{table}
